zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumZooKeeperServer.java

```diff
@@ -189,9 +189,18 @@ public void dumpConf(PrintWriter pwriter) {
         pwriter.print(self.getQuorumVerifier().toString());
     }
 
+    private final Object stateChangeMutex = new Object();
+
     @Override
     protected void setState(State state) {
-        this.state = state;
+        synchronized (stateChangeMutex) {
+            if (this.state == State.ERROR) {
+                if (state == State.RUNNING || state == State.INITIAL) {
+                    return;
```

Comment:
we should add a relevant log in order to see if this case is happening.

Creating a test that reproduces the problem and demonstrates that this fix actually resolves the problem would be better

Reply:
To add the test, do we need to consider using a fault injection tool? See [ZOOKEEPER-3601](https://issues.apache.org/jira/browse/ZOOKEEPER-3601) and https://github.com/apache/zookeeper/pull/1135

Reply:
> 100% agree with @eolivelli on adding a `LOG.warn` (if not `LOG.error`).

Okay, I will add the log later.

> For the test, I would suggest mocking and/or overriding some methods to explicitly generate the fault, if possible. I haven't tried doing so, but perhaps `LearnerTest.java` (which subclasses `LearnerZooKeeperServer`) can serve as inspiration?

Okay, I will take a try.
BTW, I think introducing a fault injection tool like https://github.com/apache/zookeeper/pull/1135 may make it easier to write a test for it, and is helpful for the community. Another bug I propose (https://github.com/apache/zookeeper/pull/1582) also require such tool to reproduce/test. I commented at https://github.com/apache/zookeeper/pull/1135 but haven't got response. Do you have any suggestion or opinion? Thank you.

Reply:
Hi @functioner, I am very interested in this problem, I have written a test case, I don't konw if it can be used to test this problem.Maybe you can use it as a reference.
[TestCase](https://github.com/lanicc/zookeeper/blob/state_byteman/zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/LeaderLeadingStateTest.java)
[Byteman btm script](https://github.com/lanicc/zookeeper/blob/state_byteman/zookeeper-server/src/test/resources/test-leader-leading-state.btm)

Reply:
if you want to use `stateChangeMutex` to handle concurrent access to `state` you have to use it at every read/write access.

Using AtomicReference is better and easier to understand.
I suggest to switch to `AtomicReference` and drop `stateChangeMutex`